apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts

```diff
@@ -0,0 +1,201 @@
+    ) {
+      setStatus('pending')
+      wrappedLoadData(account, sdk, swapCalls)
+        .then(([reqStatus, address, recievedMasterInput]) => {
+          setStatus(reqStatus as WallchainStatus)
+          setApprovalAddress(address)
+          setMasterInput(recievedMasterInput)
+        })
+        .catch((e) => {
+          setStatus('not-found')
+          setApprovalAddress(undefined)
+          setMasterInput(undefined)
+          captureException(e)
+        })
+    } else {
+      setStatus('not-found')
+      setApprovalAddress(undefined)
+      setMasterInput(undefined)
+    }
+  }, [walletClient, account, swapCalls])
+
+  return [status, approvalAddress, masterInput]
+}
+
+export function useWallchainSwapCallArguments(
+  trade: SmartRouterTrade<TradeType> | undefined | null,
+  previousSwapCalls: { address: `0x${string}`; calldata: `0x${string}`; value: `0x${string}` }[] | undefined | null,
+  account: string | undefined | null,
+  masterInput?: string,
+) {
+  const [swapCalls, setSwapCalls] = useState<SwapCall[] | WallchainSwapCall[]>([])
+  const { data: walletClient } = useWalletClient()
+
+  const [srcToken, dstToken] = extractTokensFromTrade(trade)
+  const amountIn = trade?.inputAmount?.numerator?.toString() as `0x${string}`
+  const needPermit = !trade?.inputAmount?.currency?.isNative
+
+  useEffect(() => {
+    ;(async () => {
+      if (
+        !walletClient ||
+        !masterInput ||
+        !srcToken ||
+        !dstToken ||
+        !amountIn ||
+        !previousSwapCalls ||
+        !previousSwapCalls[0] ||
+        !account
+      ) {
+        if (!previousSwapCalls || !previousSwapCalls.length) {
+          setSwapCalls([])
+        } else {
+          setSwapCalls(previousSwapCalls)
+        }
+
+        return
+      }
+
+      const callback = async () => {
+        try {
+          const sdk = new WallchainSDK({
+            keys: WallchainKeys,
+            provider: walletClient.transport as TOptions['provider'],
+            overrideAddresses,
+          })
+          const spender = (await sdk.getSpender()) as `0x${string}`
+
+          let witness: false | Awaited<ReturnType<typeof sdk.signPermit>> = false
+          if (needPermit) {
+            witness = await sdk.signPermit(srcToken as `0x${string}`, account, spender, amountIn)
+          }
+
+          const data = await sdk.createNewTransaction(
+            {
+              value: previousSwapCalls[0].value,
+              to: previousSwapCalls[0].address,
+              data: previousSwapCalls[0].calldata,
+              from: account,
+              srcToken: srcToken as `0x${string}`,
+              dstToken: dstToken as `0x${string}`,
+              amountIn,
+              isPermit: false,
+            },
+            masterInput,
+            witness,
+          )
+
+          return {
+            address: data.to as `0x${string}`,
+            calldata: data.data as `0x${string}`,
```

Comment:
can we add a checking the `data.data` match the `previousSwapCalls[0].data` here?